src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(stateClone, action.extraData);
+
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete stateClone[key];
+        }
+
+        break;
+      }
+
+      case 'clear': {
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+
+        break;
+      }
+    }
+
+    result.push({ ...stateClone });
```

Comment:
It's better to push the cloned state to the 'result' array outside of the switch block, but still inside the loop, after the current action has been processed. This ensures that the action is applied to the state copy before it is added to the result array.